libs/langchain/langchain/tools/base.py

```diff
@@ -533,17 +536,24 @@ def __init__(
     @classmethod
     def from_function(
         cls,
-        func: Callable,
+        func: Optional[Callable],
         name: str,  # We keep these required to support backwards compatibility
         description: str,
         return_direct: bool = False,
         args_schema: Optional[Type[BaseModel]] = None,
+        coroutine: Optional[
+            Callable[..., Awaitable[Any]]
+        ] = None,  # This is last for compatibility, but should be after func
         **kwargs: Any,
     ) -> Tool:
         """Initialize tool from a function."""
+        assert (
```

Comment:
Let's change to a value error instead of an `assert` statement

Reply:
While i'd also prefer an error over an assert, I was following the established style for these functions. There are already asserts there so I was just following the established practice.

But if you think we should use errors instead that's fine, it just might be a bit wonky seeing the 2 styles mixed. If you decorate an async function it will throw an error, and if you decorate a regular function you'll get a failed assert.

Reply:
Hi there thanks for your patience. We can land with assertions